src/sentry/ingest/transaction_clusterer/tree.py

```diff
@@ -0,0 +1,136 @@
+""" Build a directory tree from URL patterns and merge nodes with many siblings.
+
+For example, a blog project might contain transaction names along the lines of:
+
+/users/my-user-name/posts/2022-01-01-that-one-time-i-did-something
+
+We build a tree that looks like this:
+
+/users
+  /my-user-name
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+  /my-user-name2
+    /posts
+      /2022-12-31-i-did-something-too
+  /my-user-name3
+    /posts
+      /2022-07-15-but-what-about-me
+    /settings
+
+As soon as the node /users has more than X children, we merge them:
+
+/users
+  /*
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+      /2022-12-31-i-did-something-too
+      /2022-07-15-but-what-about-me
+    /settings
+
+If /posts now also has more than X children, those are merged as well:
+
+/users
+  /*
+    /posts
+      *
+    /settings
+
+Resulting in the following replacement rules:
+
+/users/*/posts/*  # To replace both identifiers
+/users/*          # For URLs with only one identifier, e.g. /users/*/settings
+
+"""
+
+from collections import UserDict, defaultdict
+from typing import Iterable, List, Optional, Union
+
+from typing_extensions import TypeAlias
+
+from .base import Clusterer, ReplacementRule
+
+__all__ = ["TreeClusterer"]
+
+
+class Merged:
+    pass
+
+
+#: Symbol representing a merged node
+MERGED = Merged()
+
+#: Separator by which we build the tree
+SEP = "/"
+
+
+# TODO: Metrics for tree size, spans for merge, etc.
+
+
+class TreeClusterer(Clusterer):
+    def __init__(self, *, merge_threshold: int) -> None:
+        self._merge_threshold = merge_threshold
+        self._graph = Node()
```

Comment:
Now that I'm thinking about this, it seems to me we're building our own [trie](https://en.wikipedia.org/wiki/Trie). I've looked a bit for existing modules in Python we could reuse, but haven't found anything useful.

Reply:
Good find, yeah that's exactly what this is.